operators/src/processing/expression.rs

```diff
@@ -0,0 +1,361 @@
+                    spatial_reference: SpatialReferenceOption::None, // TODO
+                })
+            },
+            self.raster_sources,
+            vec![], // TODO: is it ok to omit them?
+        )
+        .map(InitializedOperatorImpl::boxed)
+    }
+}
+
+impl InitializedOperator<RasterResultDescriptor, TypedRasterQueryProcessor>
+    for InitializedOperatorImpl<ExpressionParams, RasterResultDescriptor, ()>
+{
+    fn query_processor(&self) -> Result<TypedRasterQueryProcessor> {
+        // TODO: handle different number of sources
+
+        match *self.raster_sources.as_slice() {
+            [ref a, ref b] => {
+                let a = a.query_processor()?;
+                let b = b.query_processor()?;
+
+                call_bi_generic_processor!(a, b, (p_a, p_b) => {
+                    let res = call_generic_raster_processor!(self.params.output_type, ExpressionQueryProcessor::new(
+                                &SafeExpression {
+                                    expression: self.params.expression.clone(),
+                                },
+                                p_a,
+                                p_b,
+                            )
+                            .boxed());
+                    Ok(res)
+                })
+            }
+            _ => unimplemented!(), // TODO: handle more than two inputs
+        }
+    }
+}
+
+struct ExpressionQueryProcessor<T1, T2, TO>
+where
+    T1: Pixel,
+    T2: Pixel,
+    TO: Pixel,
+{
+    pub source_a: Box<dyn RasterQueryProcessor<RasterType = T1>>,
+    pub source_b: Box<dyn RasterQueryProcessor<RasterType = T2>>,
+    pub phantom_data: PhantomData<TO>,
+    pub cl_program: CompiledCLProgram,
+}
+
+impl<T1, T2, TO> ExpressionQueryProcessor<T1, T2, TO>
+where
+    T1: Pixel,
+    T2: Pixel,
+    TO: Pixel,
+{
+    fn new(
+        expression: &SafeExpression,
+        source_a: Box<dyn RasterQueryProcessor<RasterType = T1>>,
+        source_b: Box<dyn RasterQueryProcessor<RasterType = T2>>,
+    ) -> Self {
+        Self {
+            source_a,
+            source_b,
+            cl_program: Self::create_cl_program(&expression),
+            phantom_data: PhantomData::default(),
+        }
+    }
+
+    fn create_cl_program(expression: &SafeExpression) -> CompiledCLProgram {
+        // TODO: generate code for arbitrary amount of inputs
+        let source = r#"
+__kernel void expressionkernel(
+            __global const IN_TYPE0 *in_data0,
+            __global const RasterInfo *in_info0,
+            __global const IN_TYPE1* in_data1,
+            __global const RasterInfo *in_info1,
+            __global OUT_TYPE0* out_data,
+            __global const RasterInfo *out_info)
+{
+    uint const gid = get_global_id(0) + get_global_id(1) * in_info0->size[0];
+    if (gid >= in_info0->size[0]*in_info0->size[1]*in_info0->size[2])
+        return;
+
+    IN_TYPE0 A = in_data0[gid];
+    if (ISNODATA0(A, in_info0)) {
+        out_data[gid] = out_info->no_data;
```

Comment:
What if there is no NODATA in the output raster? Do we have a convention here?

Reply:
if there is no nodata set then its just some number?

Reply:
I guess the expression operator should check that if at least one input has a no data value, then the output data needs to have a no data value defined. If not, the operator should throw an error. What do you think?

Reply:
Mapping had the problem to not being able to use rasters without NODATA since it does, what Johannes suggests, just set it to some number. If the range is exhausted, e.g. for a byte raster, this becomes a problem.
So either we disallow rasters without NODATA at all or we implement Michaels approach.

Reply:
then just force a NoData value and increase the datatype if needed. Error if no nodata is possible is ok i guess...

Reply:
right now this value is 0. I think it would be better to force the output to have nodata if at least one input has no data. but right now we cant check if the input has no data before actually getting the first tile.

Reply:
Trying to create NODATA is okay I guess but isn't it strange to magically change the data type?

Reply:
Currently our rasters only have a data type and not min/max values right? Mapping uses the unit info about min/max to insert a no data value. we are not there yet. I am also against automatically increasing the data type. The user should say the desired output data type and the system should tell him to set a no data value if needed. then she can decide whether to choose a bigger data type to get a value. But this cannot be solved right now because of the lack of meta info in raster data and also has nothing to do with opencl

Reply:
yes magic type changes should be avoided

Reply:
no data is now mandatory 3fe3ae0f1429bfac027c6091a1b6c8ac2638272c